Code Break: Can set visible after for lesson

dmcavoy opened the pull request:
Part 1 of [PLAT-45](https://codedotorg.atlassian.net/browse/PLAT-45)

Levelbuilders can set `visible_after` on a stage in the script edit page. If the levelbuilder leaves the value for `visible_after` blank (`''`) then it will default to the next Wednesday at 8am PST. The value will be saved in the `.script` file and loaded onto the script edit page the next time the level is edited.

In addition when you go to the script overview page as a levelbuilder you will see a flash notice that lets you know when a lesson will be visible after.

![Screen Shot 2020-03-27 at 4 06 41 PM](https://user-images.githubusercontent.com/208083/77796036-f6dd3280-7044-11ea-9b4d-a669cc00f39f.png)

Follow up work: Next up is using this value to determine if the stage is hidden. 

## Links

<!--
  Any relevant links to external resources; ie, specification documents, jira
  items, related PRs, honeybadger errors, etc
-->

- [PLAT-45](https://codedotorg.atlassian.net/browse/PLAT-45)

## Testing story

<!--
  Does your change include appropriate tests?

  If so, please describe how the tests included in this PR are sufficient

  If not, please explain why this change does not need to be tested.
-->

Added tests to script_dsl_test.rb for parsing the DSL and serializing stage. 

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

davidsbailey commented on dashboard/app/dsl/script_dsl.rb:
```diff
@@ -82,19 +83,30 @@ def stage(name, properties = {})
+  def determine_visible_after_time(visible_after_value)
+    current_time = Time.now
+    raw_diff_to_wed = 3 - current_time.wday
+    diff_to_wed = raw_diff_to_wed > 0 ? raw_diff_to_wed : 7 + raw_diff_to_wed
```
nit: I think you can just use `raw_diff_to_wed % 7` here, because it will return a non-negative number even when the input is negative:
```
irb(main):001:0> 10 % 7
=> 3
irb(main):002:0> 3 % 7
=> 3
irb(main):003:0> -4 % 7
=> 3
```

davidsbailey commented on dashboard/app/dsl/script_dsl.rb:
```diff
@@ -82,19 +83,30 @@ def stage(name, properties = {})
+    diff_to_wed = raw_diff_to_wed > 0 ? raw_diff_to_wed : 7 + raw_diff_to_wed
+    next_wednesday = current_time + diff_to_wed.day
+    next_wednesday_8_am_pst = Time.new(next_wednesday.year, next_wednesday.month, next_wednesday.day, 8, 0, 0, '-07:00').to_s
+    visible_after_value == '' ? next_wednesday_8_am_pst : visible_after_value
```
the variable name says PST, but the offset is `-07:00` which is PDT. These should definitely be made consistent. I'd vote for sticking with PDT and just updating your variable name, so you don't have to change it in too many places, and also because that will be less unpleasant surprise in the unfortunate situation that we're still making new lessons like this in November.

davidsbailey commented on dashboard/app/controllers/scripts_controller.rb:
```diff
@@ -45,6 +45,16 @@ def show
+
+        formatted_time = Time.parse(stage.visible_after).strftime("%I:%M %p %A %B %d %Y %Z")
+        num_days_away = ((Time.parse(stage.visible_after) - Time.now) / 1.day).ceil.to_s
+        flash[:notice] = "The lesson #{stage.name} will be visible after #{formatted_time} (#{num_days_away} Days)"
```
please also test the following scenarios:
1. the banner is not visible when `visible_after` is absent
2. the banner is not visible if the time indicated by `visible_after` has passed (this may be easier to implement after you've added the code to actually hide the lesson, so ok to do in a later PR)
3. we should do something reasonable if 2 lessons are hidden. if the current code means that both banners appear, that's great. but if one replaces the other, then we might need to do something different -- anything that will prevent a nasty surprise where a lesson remains hidden when we need it to be visible.
4. no banner is visible for students

davidsbailey replied:
also, what happens if two lessons are hidden -- do we show two banners?

uponthesun commented on dashboard/app/controllers/scripts_controller.rb:
```diff
@@ -45,6 +45,16 @@ def show
     @sections_with_assigned_info = sections&.map {|section| section.attributes.merge!({"isAssigned" => section[:script_id] == @script.id})}
+
+    if current_user.levelbuilder?
+      @script.stages.each do |stage|
```
What's the intended behavior here if more than one stage has a `visible_after`?

dmcavoy replied:
@davidsbailey what was your thoughts here on intended behavior

dmcavoy replied:
It now concats multiple warnings together

dmcavoy commented:
Updated the flash notice to concat together when multiple lessons have visible after set.

<img width="830" alt="Screen Shot 2020-03-31 at 3 51 54 PM" src="https://user-images.githubusercontent.com/208083/78069107-8ea19000-7367-11ea-8378-678921496b5e.png">

Added tests for:
1. notice does not show if no visible after and levelbuilder
2. notice does not show if visible after date in past and levelbuilder
3. notice shows if visible after date is in future and levelbuilder
4. notice does not show for visible after if student
5. notice does not show for visible after if teacher
